crates/sui-json-rpc/src/name_service.rs

```diff
@@ -154,12 +160,59 @@ impl FromStr for Domain {
             return Err(DomainParseError::LabelsEmpty);
         }
 
+        add_tld_if_missing(&mut labels);
+
         let labels = labels.into_iter().map(ToOwned::to_owned).collect();
 
         Ok(Domain { labels })
     }
 }
 
+/// Parses a separator from the domain string input.
+/// E.g.  `example.sui` -> `.` | example@sui -> `@` | `example*sui` -> `*`
+fn separator(s: &str) -> Result<char, DomainParseError> {
+    let mut domain_separator: Option<char> = None;
+
+    for separator in ACCEPTED_SEPARATORS.iter() {
+        if s.contains(*separator) {
+            if domain_separator.is_some() {
+                return Err(DomainParseError::InvalidSeparator);
+            }
+
+            domain_separator = Some(*separator);
+        }
+    }
+
+    match domain_separator {
+        Some(separator) => Ok(separator),
+        None => Ok(ACCEPTED_SEPARATORS[0]),
+    }
+}
+
+/// Removes separator from the beggining of the input
+/// E.g. `@example` -> `example` | `@example@sui` -> `example@sui`
+fn remove_leading_separator(s: &str, separator: char) -> &str {
+    if let Some(c) = s.chars().next() {
+        if c == separator {
+            return &s[1..];
+        }
+    }
+    s
```

Comment:
This implementation is a little bit dicey because `chars()` iterates over unicode scalar values, and indexing into the string indexes by byte offset.  There is not a problem today, because all your separators are single-byte unicode scalars, but to be air tight, I would do this:

```suggestion
    let mut cs = s.chars().peekable();
    if matches!(cs.peek(), Some(c) if c == separator) {
      cs.next();
    }
    cs.as_str()
```